mesonbuild/compilers/detect.py

```diff
@@ -0,0 +1,1220 @@
+from .objc import (
+    ObjCCompiler,
+    AppleClangObjCCompiler,
+    ClangObjCCompiler,
+    GnuObjCCompiler,
+)
+from .objcpp import (
+    ObjCPPCompiler,
+    AppleClangObjCPPCompiler,
+    ClangObjCPPCompiler,
+    GnuObjCPPCompiler,
+)
+from .cython import CythonCompiler
+from .rust import RustCompiler
+from .swift import SwiftCompiler
+from .vala import ValaCompiler
+from .mixins.visualstudio import VisualStudioLikeCompiler
+from .mixins.gnu import GnuCompiler
+from .mixins.clang import ClangCompiler
+
+import subprocess
+import platform
+import re
+import shutil
+import tempfile
+import os
+import typing as T
+
+if T.TYPE_CHECKING:
+    from ..environment import Environment
+    from ..programs import ExternalProgram
+    from .compilers import CompilerType
+
+
+
+# Default compilers and linkers
+# =============================
+
+defaults: T.Dict[str, T.List[str]] = {}
+
+# List of potential compilers.
+if is_windows():
+    # Intel C and C++ compiler is icl on Windows, but icc and icpc elsewhere.
+    # Search for icl before cl, since Intel "helpfully" provides a
+    # cl.exe that returns *exactly the same thing* that microsofts
+    # cl.exe does, and if icl is present, it's almost certainly what
+    # you want.
+    defaults['c'] = ['icl', 'cl', 'cc', 'gcc', 'clang', 'clang-cl', 'pgcc']
+    # There is currently no pgc++ for Windows, only for  Mac and Linux.
+    defaults['cpp'] = ['icl', 'cl', 'c++', 'g++', 'clang++', 'clang-cl']
+    defaults['fortran'] = ['ifort', 'gfortran', 'flang', 'pgfortran', 'g95']
+    # Clang and clang++ are valid, but currently unsupported.
+    defaults['objc'] = ['cc', 'gcc']
+    defaults['objcpp'] = ['c++', 'g++']
+    defaults['cs'] = ['csc', 'mcs']
+else:
+    if platform.machine().lower() == 'e2k':
+        # There are no objc or objc++ compilers for Elbrus,
+        # and there's no clang which can build binaries for host.
+        defaults['c'] = ['cc', 'gcc', 'lcc']
+        defaults['cpp'] = ['c++', 'g++', 'l++']
+        defaults['objc'] = []
+        defaults['objcpp'] = []
+    else:
+        defaults['c'] = ['cc', 'gcc', 'clang', 'nvc', 'pgcc', 'icc']
+        defaults['cpp'] = ['c++', 'g++', 'clang++', 'nvc++', 'pgc++', 'icpc']
+        defaults['objc'] = ['cc', 'gcc', 'clang']
+        defaults['objcpp'] = ['c++', 'g++', 'clang++']
+    defaults['fortran'] = ['gfortran', 'flang', 'nvfortran', 'pgfortran', 'ifort', 'g95']
+    defaults['cs'] = ['mcs', 'csc']
+defaults['d'] = ['ldc2', 'ldc', 'gdc', 'dmd']
+defaults['java'] = ['javac']
+defaults['cuda'] = ['nvcc']
+defaults['rust'] = ['rustc']
+defaults['swift'] = ['swiftc']
+defaults['vala'] = ['valac']
+defaults['cython'] = ['cython']
+defaults['static_linker'] = ['ar', 'gar']
+defaults['strip'] = ['strip']
+defaults['vs_static_linker'] = ['lib']
+defaults['clang_cl_static_linker'] = ['llvm-lib']
+defaults['cuda_static_linker'] = ['nvlink']
+defaults['gcc_static_linker'] = ['gcc-ar']
+defaults['clang_static_linker'] = ['llvm-ar']
```

Comment:
I konw this is kinda syle-ish, buut maybe:
```python
defaults = {
  'd': ['ldc2', 'ldc', 'gdc', 'dmd'],
  'java': ['javac'],
  ...
}
if mesonlib.is_windows():
  defaults.update({
    'c': ['icl', 'cl', ...],
   ...
  })
```

Reply:
I will leave it like it currently is, I am split on which option is better with the current amount of branches...